beam-college-dataflow/codelab.md

```diff
@@ -0,0 +1,356 @@
+
+```java
+void parseJson(String jsonSchema) throws UnsupportedOperationException {
+    TableSchema schema = BigQueryHelpers.fromJsonString(jsonSchema, TableSchema.class);
+    validateSchemaTypes(schema);
+    bigQuerySchema = schema;
+    jsonBeamSchema = BigQueryHelpers.toJsonString(schema.getFields());
+}
+```
+
+And convert ```TableSchema``` to Beam `Schema` using `org.apache.beam.sdk.io.gcp.bigquery.BigQueryUtils.fromTableSchema`
+
+You may see how it implements in presented pipeline `v2/protegrity-data-tokenization/src/main/java/com/google/cloud/teleport/v2/utils/SchemasUtils.java` class
+
+Nex step: IO transform overview
+<!-- ------------------------ -->
+## GcsIO overview
+Duration: 0:10:00
+
+To transform many input formats to the Beam Row and from Beam Row into destination format, presented
+pipeline implements IO transforms. Let's focus on GcsIO that works with Google Cloud Storage.
+
+In the `v2/protegrity-data-tokenization/src/main/java/com/google/cloud/teleport/v2/transforms/io/GcsIO.java` you may see `FORMAT` enum, that covers all supported formats.
+```java
+public enum FORMAT {
+    JSON,
+    CSV,
+    AVRO
+}
+```
+
+For IO functionality `GcsIO` class provides two methods `read` 
+```java
+public PCollection<Row> read(Pipeline pipeline, SchemasUtils schema) {
+    switch (options.getInputGcsFileFormat()) {
+      case CSV:
+        ...
+      case JSON:
+        return readJson(pipeline)
+            .apply(new JsonToBeamRow(options.getNonTokenizedDeadLetterGcsPath(), schema));
+      case AVRO:
+        return readAvro(pipeline, schema.getBeamSchema());
+      default:
+        throw new IllegalStateException(
+            "No valid format for input data is provided. Please, choose JSON or CSV.");
+
+    }
+  }
+```
+
+and `write`
+```java
+public POutput write(PCollection<Row> input, Schema schema) {
+    switch (options.getOutputGcsFileFormat()) {
+      case JSON:
+        return writeJson(input);
+      case AVRO:
+        return writeAvro(input, schema);
+      case CSV:
+        return writeCsv(input, schema.getFieldNames());
+      default:
+        throw new IllegalStateException(
+            "No valid format for output data is provided. Please, choose JSON or CSV.");
+
+    }
+}
+```
+
+For each supported format GcsIO encapsulates transformations from source format to Beam Row and 
+from Beam Row to destination format for writing results. e.g. `writeJson` or `readJson`. 
+That takes pipeline and Schema as argument.
+
+Next step: You will implement parquet files reading and writing and transformation this into the Beam Row.
+<!-- ------------------------ -->
+## Parquet IO Read
+Duration: 0:15:00
+
+To implement new parquet source read and write, firstly you need to add new value into `FORMAT` enum 
+
+```java
+public enum FORMAT {
+    JSON,
+    CSV,
+    AVRO,
+    PARQUET
+  }
+```
+
+Next you need to create method for read parquet from filesystem and convert it into Beam Row.
```

Comment:
Next, you need to create a method to read parquet from the filesystem and convert it into Beam Row.